With.Fody/ModuleWeaver.cs

```diff
@@ -20,52 +21,57 @@ public ModuleWeaver()
 
     public void Execute()
     {
-        foreach (var type in ModuleDefinition.Types.Where(CanHaveWith))
+        foreach (var type in ModuleDefinition.Types)
         {
-            AddWith(type);
-            RemoveGenericWith(type);
-            LogInfo($"Added method 'With' to type '{type.Name}'.");
+            if (type.GetMethods().Any(m => m.IsPublic && m.Name.StartsWith("With")))
+            {
+                var ctor = GetValidConstructor(type);
+                if (ctor != null)
+                {
+                    AddWith(type, ctor);
+                    RemoveGenericWith(type);
+                    LogInfo($"Added method 'With' to type '{type.Name}'.");
+                }
+            }
         }
     }
 
-    private bool CanHaveWith(TypeDefinition type)
+    private static MethodDefinition GetValidConstructor(TypeDefinition type)
     {
-        if (!type.GetMethods().Any(m => m.IsPublic && m.Name.StartsWith("With")))
-        {
-            return false;
-        }
-
-        var ctors = type.GetConstructors().Where(c => c.IsPublic).ToArray();
-        if (ctors.Length != 1)
-        {
-            return false;
-        }
-
-        var ctor = ctors[0];
-        var parameters = ctor.Parameters;
-        if (parameters.Count < 2)
-        {
-            return false;
-        }
-
-        return parameters.All(par => type.Properties.Any(pro => string.Compare(par.Name, pro.Name, StringComparison.InvariantCultureIgnoreCase) == 0));
+        return type.GetConstructors()
+            .Where(ctor =>
+                ctor.Parameters.Count >= 2 &&
+                ctor.Parameters.All(par =>
+                    type.Properties.Any(pro =>
+                        string.Compare(par.Name, pro.Name, StringComparison.InvariantCultureIgnoreCase) == 0
+                    )
+                )
+            )
+            .FirstOrDefault();
```

Comment:
This query looks for constructors with the exact number of parameters as properties. For the case you mention, only the 5 parameter constructor would be returned.
The only case where more that one constructor would be found is if they had the same number of parameters but in a different order. I've never seen this case in the real world. It would be strange to have constructors behaving differently based on the order of parameters.
The query also checks for the names and I'm now adding type checking.
_With.Fody_ should allow any number of constructors as long as there is one that can be injected into the _.With_ methods. _GetValidConstructor()_ finds exactly that one, ignoring all the others.
I'm using _With.Fody_ on a large project and some more complex scenarios appear. Thanks again for developing this module as it's saving me a lot of trouble.